op-conductor/rpc/ws/server.go

```diff
@@ -0,0 +1,231 @@
+			}
+			if dropCount > 0 {
+				h.log.Warn("Failed to send message to all clients, dropped", "successCount", successCount, "dropCount", dropCount)
+			}
+		}
+	}
+}
+
+// Client represents a connected WebSocket client
+type Client struct {
+	conn   *websocket.Conn
+	send   chan []byte
+	ctx    context.Context
+	cancel context.CancelFunc
+	hub    *Hub
+	log    log.Logger
+	mu     sync.Mutex
+}
+
+// Close closes the client's WebSocket connection and send channel
+func (c *Client) Close() {
+	// this mutex is used to prevent concurrent close operations
+	// double close will panic
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.cancel()
+	c.conn.Close(websocket.StatusNormalClosure, "client closed")
+	close(c.send)
+}
+
+// newClient creates a new client with default settings
+func newClient(conn *websocket.Conn, ctx context.Context, hub *Hub, logger log.Logger) *Client {
+	ctx, cancel := context.WithCancel(ctx)
+	return &Client{
+		conn:   conn,
+		send:   make(chan []byte, sendChannelBufferSize),
+		ctx:    ctx,
+		cancel: cancel,
+		hub:    hub,
+		log:    logger,
+	}
+}
+
+// serveWs handles WebSocket requests from clients
+func (h *Handler) serveWs(w http.ResponseWriter, r *http.Request) {
+	// Upgrade the HTTP connection to a WebSocket connection using coder/websocket
+	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
+		CompressionMode: websocket.CompressionDisabled,
+	})
+
+	if err != nil {
+		h.log.Error("failed to upgrade connection", "err", err)
+		return
+	}
+
+	client := newClient(conn, r.Context(), h.hub, h.log)
+
+	// Register the client with the hub
+	h.hub.register <- client
+	h.log.Info("WebSocket client connected")
+
+	// Start client handling
+	go h.writePump(client)
+	h.readPump(client)
+}
+
+// readPump for followers where you don't expect regular data messages
+func (h *Handler) readPump(client *Client) {
+	defer func() {
+		// Unregister the client when the read pump exits
+		h.hub.unregister <- client
+		h.log.Info("WebSocket read pump exited, client unregistered")
+	}()
+
+	for {
+		select {
+		case <-client.ctx.Done():
+			return
+		default:
+			// Always read to process control frames (ping/pong/close)
+			readCtx, _ := context.WithTimeout(client.ctx, 30*time.Second)
+			messageType, message, err := client.conn.Read(readCtx)
+
+			if err != nil {
+				if errors.Is(err, context.DeadlineExceeded) {
+					// Timeout is expected when no messages - continue reading
+					h.log.Debug("Read timeout occurred, continuing to process control frames")
+					continue
+				}
+				if websocket.CloseStatus(err) != -1 {
+					h.log.Debug("Client closed connection", "code", websocket.CloseStatus(err))
+					return
+				}
+				h.log.Debug("Error reading from WebSocket client", "err", err)
+				return
+			}
+
+			// Handle any data messages from clients if needed
+			if messageType == websocket.MessageText || messageType == websocket.MessageBinary {
+				h.log.Debug("Received message from client", "message", string(message))
+			}
```

Comment:
```suggestion
		h.log.Debug("Received message from client", "message", string(message))
```

Glancing over the underlying implementation, seems like the messageType will only either be Text or Binary.
- Ping, Pong, Close opcodes ->  get stuck in a read loop of control frames.
- Continuation opcode and any other unknown other op-code -> returns an error